pkg/templates/generic_lookup_func.go

```diff
@@ -297,3 +299,109 @@ func onAllowlist(allowlist []ClusterScopedObjectIdentifier, rsrc ClusterScopedOb
 
 	return false
 }
+
+func (t *TemplateResolver) getNodesWithRoleHelper(
+	options *ResolveOptions,
+	templateResult *TemplateResult,
+) func(...string) (
+	map[string]interface{}, error,
+) {
+	return func(name ...string) (
+		map[string]interface{}, error,
+	) {
+		return t.getNodesWithRole(options, templateResult, name...)
+	}
+}
+
+// function getNodesWithRole gets a listing of all nodes with only the
+// role specified.  Any nodes that include other roles in addition
+// to the specified roles are not included.
+func (t *TemplateResolver) getNodesWithRole(
+	options *ResolveOptions,
+	templateResult *TemplateResult,
+	name ...string,
+) (
+	map[string]interface{}, error,
+) {
+	result := []unstructured.Unstructured{}
+
+	if len(name) == 0 || strings.TrimSpace(name[0]) == "" {
+		return nil, fmt.Errorf("%w: name must be specified", ErrInvalidInput)
+	}
+
+	var searchRoles []string
+
+	for _, n := range name {
+		searchRoles = append(searchRoles, fmt.Sprintf("node-role.kubernetes.io/%v", n))
+	}
+
+	nodes, err := t.getOrList(options, templateResult, "v1", "Node", "", "", searchRoles...)
+	if err != nil {
+		return nil, err
+	}
+
+	var nodeList unstructured.UnstructuredList
+
+	nlByte, err := json.Marshal(nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	_ = json.Unmarshal(nlByte, &nodeList)
+
+	// check if the nodes found contain any other role labels
+	for _, node := range nodeList.Items {
+		isMatchedNode := true
+
+		for k := range node.GetLabels() {
+			if strings.HasPrefix(k, "node-role.kubernetes.io") &&
+				!slices.Contains(searchRoles, k) &&
+				k != "node-role.kubernetes.io/worker" {
+				isMatchedNode = false
+
+				break
+			}
+		}
+
+		if isMatchedNode {
+			result = append(result, node)
+		}
+	}
+
+	resultList := unstructured.UnstructuredList{Items: result}
+
+	return resultList.UnstructuredContent(), nil
+}
+
+func (t *TemplateResolver) hasNodesWithRoleHelper(
+	options *ResolveOptions,
+) func(...string) (bool, error) {
+	return func(name ...string) (
+		bool, error,
+	) {
+		return t.hasNodesWithRole(options, name...)
+	}
+}
+
+// function hasNodesWithRole returns true if there are any nodes labeled with only the
+// specified role.  Does not include nodes which have additional roles on them.
+func (t *TemplateResolver) hasNodesWithRole(
+	options *ResolveOptions, name ...string,
+) (bool, error) {
+	var resolvedResult TemplateResult
+	var nodeList unstructured.UnstructuredList
+
+	nodes, err := t.getNodesWithRole(options, &resolvedResult, name...)
+	if err != nil {
+		return false, err
+	}
+
+	nlByte, err := json.Marshal(nodes)
+	if err != nil {
+		return false, err
+	}
+
+	_ = json.Unmarshal(nlByte, &nodeList)
+
+	return (len(nodeList.Items) > 0), nil
```

Comment:
Does this work to prevent the marhsal and unmarshal?
```suggestion
	items, _, _ := unstructured.NestedSlice(nodes, "items")

	return len(items) > 0, nil
```

Reply:
agree